python/src/com/jetbrains/python/psi/impl/PyTargetExpressionImpl.java

```diff
@@ -243,24 +245,35 @@ private static PyType getWithItemVariableType(TypeEvalContext context, PyWithIte
     final PyExpression expression = item.getExpression();
     if (expression != null) {
       final PyType exprType = context.getType(expression);
-      if (exprType instanceof PyClassType) {
-        final PyClass cls = ((PyClassType)exprType).getPyClass();
-        final PyFunction enter = cls.findMethodByName(PyNames.ENTER, true, null);
-        if (enter != null) {
-          final PyType enterType = enter.getCallType(expression, Collections.emptyMap(), context);
-          if (enterType != null) {
-            return enterType;
-          }
-          for (PyTypeProvider provider : Extensions.getExtensions(PyTypeProvider.EP_NAME)) {
-            PyType typeFromProvider = provider.getContextManagerVariableType(cls, expression, context);
-            if (typeFromProvider != null) {
-              return typeFromProvider;
-            }
-          }
-          // Guess the return type of __enter__
-          return PyUnionType.createWeakType(exprType);
+      if (exprType instanceof PyClass) {
+        return getEnterTypeFromPyClass(context, expression, (PyClassType)exprType);
+      }
+      else if (exprType instanceof PyUnionType) {
+        List<PyType> collect = StreamEx.of(((PyUnionType)exprType).getMembers())
+          .select(PyClassType.class)
+          .map(t -> getEnterTypeFromPyClass(context, expression, t)).collect(Collectors.toList());
```

Comment:
With `StreamEx` library it's possible to use `toList()` instead of `collect(Collectors.toList())`